### PR TITLE
Use ReadView in RPC handlers.

### PR DESCRIPTION
**Pull request by rec**
@vinniefalco

### CODE REVIEW

File: src/ripple/app/ledger/Ledger.cpp [hunk 1]
```diff
@@ -246,31 +234,42 @@ Ledger::Ledger (bool /* dummy */,
...
 Ledger::Ledger (void const* data,
         std::size_t size, bool hasPrefix)
+
     : mImmutable (true)
+    , txMap_ (std::make_shared <SHAMap> (
+          SHAMapType::TRANSACTION, getApp().family(),
+            deprecatedLogs().journal("SHAMap")))
+    , stateMap_ (std::make_shared <SHAMap> (
+          SHAMapType::STATE, getApp().family(),
+            deprecatedLogs().journal("SHAMap")))
```
Comment (vinniefalco) [inline src/ripple/app/ledger/Ledger.cpp]: Were these initializers previously missing?

File: src/ripple/app/ledger/Ledger.h [hunk 2]
```diff
@@ -258,57 +242,9 @@ class Ledger
...
-        return mAccountHash;
-    }
-
-    std::uint64_t getTotalCoins () const
-    {
-        return mTotCoins;
-    }
-
-    void setTotalCoins (std::uint64_t totCoins)
+    void setTotalCoins (std::uint64_t totDrops)
```
Comment (vinniefalco) [inline src/ripple/app/ledger/Ledger.h]: it'd be nice to do away with this function, we might not need to set total coins any more. We should check callers. Perhaps best in a separate PR

File: src/ripple/app/ledger/impl/LedgerToJson.cpp [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/app/ledger/LedgerToJson.h>
```
Comment (vinniefalco) [inline src/ripple/app/ledger/impl/LedgerToJson.cpp]: Missing `<BeastConfig.h>`

File: src/ripple/app/misc/NetworkOPs.cpp [hunk 4]
```diff
@@ -2911,7 +2913,7 @@ void NetworkOPsImp::getBookPage (
 // FIXME : support iLimit.
 void NetworkOPsImp::getBookPage (
     bool bAdmin,
-    Ledger::pointer lpLedger,
+    std::shared_ptr<ReadView const>& lpLedger,
```
Comment (vinniefalco) [inline src/ripple/app/misc/NetworkOPs.cpp]: The type changed here, previous was passed by value and the current is passed by non-const reference. Was this intended?

File: src/ripple/ledger/ReadView.h [hunk 5]
```diff
@@ -73,12 +73,24 @@ struct LedgerInfo
...
-    //std::uint64_t coins = 0;
-    //bool validated = false;
-    //int closeTimeRes = 0;
+    std::uint64_t drops = 0;
+
+    // If validated is false, it means "not yet validated."
+    // Once validated is true, it will never be set false at a later time.
+    mutable
+    bool validated = false;
+    bool accepted = false;
```
Comment (vinniefalco) [inline src/ripple/ledger/ReadView.h]: `accepted` is deprecated, we should strive to not include it in `LedgerInfo`

File: src/ripple/ledger/ReadView.h [hunk 6]
```diff
@@ -285,8 +310,28 @@ class DigestAwareReadView
     digest (key_type const& key) const = 0;
 };
 
-// For now
-using BasicView = ReadView;
+/** Run a functor on each SLE in a ReadView starting from the key start,
+    as long as the functor returns true.
+ */
+template <class Functor>
+void forEachSLE(ReadView const& view, Functor func, uint256 const& start = {})
```
Comment (vinniefalco) [inline src/ripple/ledger/ReadView.h]: This is insanity, it visits every SLE! It should not be part of a public interface. If we need it somewhere, put the declaration and definition in the translation unit that uses it.

File: src/ripple/ledger/ReadView.h [hunk 7]
```diff
@@ -285,8 +310,28 @@ class DigestAwareReadView
...
+    for (auto k = view.succ(start); k; k = view.succ(*k))
+        if (auto sle = view.read(keylet::unchecked(*k)))
+            if (! func(*sle))
+                break;
+}
+
+/** Run a functor on each transaction in a ReadView, as long as the functor
+    returns true. Might throw an exception if the ledger is corrupt. */
+template <class Functor>
+void forEachTx(ReadView const& view, Functor func)
```
Comment (vinniefalco) [inline src/ripple/ledger/ReadView.h]: Please make this part of an implementation and not a public function

File: src/ripple/ledger/ReadView.h [hunk 8]
```diff
@@ -285,8 +310,28 @@ class DigestAwareReadView
...
+{
+    for (auto i = view.txs.begin(); i != view.txs.begin(); ++i)
+        if (i->first)
```
Comment (vinniefalco) [inline src/ripple/ledger/ReadView.h]: `i->first` can never be nullptr, its an invariant

File: src/ripple/rpc/handlers/AccountInfo.cpp [hunk 9]
```diff
@@ -58,16 +58,16 @@ Json::Value doAccountInfo (RPC::Context& context)
     if (jvAccepted)
         return jvAccepted;
 
-    auto const sleAccepted = cachedRead(*ledger,
-        keylet::account(accountID).key, ltACCOUNT_ROOT);
-
+    auto const sleAccepted = ledger->read(
+        {ltACCOUNT_ROOT, keylet::account(accountID).key});
```
Comment (vinniefalco) [inline src/ripple/rpc/handlers/AccountInfo.cpp]: `ledger->read(keylet::account(accountID))`

File: src/ripple/rpc/handlers/AccountInfo.cpp [hunk 10]
```diff
@@ -58,16 +58,16 @@ Json::Value doAccountInfo (RPC::Context& context)
...
     if (sleAccepted)
     {
         injectSLE(jvAccepted, *sleAccepted);
 
         // See if there's a SignerEntries for this account.
         uint256 const signerListIndex = getSignerListIndex (accountID);
-        auto const signerList = cachedRead(*ledger, signerListIndex);
+        auto const signerList =
+                ledger->read (keylet::unchecked(signerListIndex));
```
Comment (vinniefalco) [inline src/ripple/rpc/handlers/AccountInfo.cpp]: `ledger->read(keylet::signers(accountID))`

File: src/ripple/rpc/impl/GetAccountObjects.cpp [hunk 11]
```diff
@@ -55,17 +55,17 @@ getAccountObjects (Ledger const& ledger, AccountID const& account,
...
                 return false;
-            
+
             found = true;
         }
 
         for (; iter != entries.end (); ++iter)
         {
-            auto const sleNode = cachedRead(ledger, *iter );
+            auto const sleNode = ledger.read(keylet::unchecked(*iter));
```
Comment (vinniefalco) [inline src/ripple/rpc/impl/GetAccountObjects.cpp]: If we're iterating the children of a directory then the correct `Keylet` is `keylet::child(*iter)`

File: src/ripple/rpc/impl/LookupLedger.cpp [hunk 12]
```diff
@@ -181,38 +151,68 @@ bool isValidated (Ledger& ledger)
 // return value.  Otherwise, the object contains the field "validated" and
 // optionally the fields "ledger_hash", "ledger_index" and
 // "ledger_current_index", if they are defined.
+Status lookupLedgerDeprecated (
+    Ledger::pointer& ledger, Context& context, Json::Value& result)
+{
+    if (auto status = ledgerFromRequest (ledger, context))
```
Comment (vinniefalco) [inline src/ripple/rpc/impl/LookupLedger.cpp]: `Status` has a conversion from `bool`? What does `true` mean?

File: src/ripple/rpc/impl/LookupLedger.cpp [hunk 13]
```diff
@@ -34,15 +35,16 @@ bool isValidatedOld ()
         Tuning::maxValidatedLedgerAge;
 }
 
-Status ledgerFromRequest (
-    Json::Value const& params,
-    Ledger::pointer& ledger,
-    NetworkOPs& netOps)
+template <class T>
+Status ledgerFromRequest (T& ledger, Context& context)
```
Comment (vinniefalco) [inline src/ripple/rpc/impl/LookupLedger.cpp]: What types of `T` is this called with?

Comment (vinniefalco) [inline src/ripple/rpc/impl/LookupLedger.cpp]: I suspect a better signature is
`Status ledgerFromRequest (ReadView const& ledger, Context& context)`

**Comment by rec** [comment]
Pushed a huge change to split up the RPC handlers for a classic build...

File: src/ripple/app/misc/NetworkOPs.cpp (hunk 4 above)
Comment (rec) [inline src/ripple/app/misc/NetworkOPs.cpp]: Sorry for the mistake - the code is not compiled so such errors get ignored.

We should really get rid of these, they've been here for over a year.

Comment (vinniefalco) [inline src/ripple/app/misc/NetworkOPs.cpp]: If its not called we should delete it

Comment (rec) [inline src/ripple/app/misc/NetworkOPs.cpp]: I think @nbougalis is the one who wants to keep this.

**Comment by vinniefalco** [comment]
vcxproj out of date

File: src/ripple/ledger/ReadView.h (hunk 7 above)
Comment (rec) [inline src/ripple/ledger/ReadView.h]: Don't we prefer non-friend, non-member functions?

Comment (vinniefalco) [inline src/ripple/ledger/ReadView.h]: Yes. I'm not suggesting to make it part of a class, just to move it to the .cpp where its used.

File: src/ripple/ledger/ReadView.h (hunk 6 above)
Comment (rec) [inline src/ripple/ledger/ReadView.h]: It only visits every SLE start is empty AND if the functor returns true each time.

We use this in two separate handlers but we'll end up using this every time we want to iterate a subrange of the SLEs.  Given that iterating is a non-trivial block of code...

File: src/ripple/ledger/ReadView.h (hunk 8 above)
Comment (rec) [inline src/ripple/ledger/ReadView.h]: Done!

File: src/ripple/rpc/handlers/AccountInfo.cpp (hunk 10 above)
Comment (rec) [inline src/ripple/rpc/handlers/AccountInfo.cpp]: Done.

File: src/ripple/rpc/handlers/AccountInfo.cpp (hunk 9 above)
Comment (rec) [inline src/ripple/rpc/handlers/AccountInfo.cpp]: Done.

File: src/ripple/rpc/impl/LookupLedger.cpp (hunk 12 above)
Comment (rec) [inline src/ripple/rpc/impl/LookupLedger.cpp]: `true` is  `state != Status::OK`.

File: src/ripple/ledger/ReadView.h (hunk 6 above)
Comment (vinniefalco) [inline src/ripple/ledger/ReadView.h]: Can the declaration be narrowed just to the routines that use it?

File: src/ripple/rpc/impl/LookupLedger.cpp (hunk 13 above)
Comment (rec) [inline src/ripple/rpc/impl/LookupLedger.cpp]: Done!

**Comment by vinniefalco** [comment]
:+1: This needs to get in asap

File: src/ripple/rpc/impl/LookupLedger.cpp (hunk 13 above)
Comment (rec) [inline src/ripple/rpc/impl/LookupLedger.cpp]: Nope.  It's assigning into a _shared_ptr_ reference, so it needs to be a template (for shortest code).  It'll all go soon, we only have two more uses left of the deprecated version.

**Comment by rec** [comment]
I can squash this in seconds... :-D

**Comment by vinniefalco** [comment]
`JUMP MISMATCH` is bad

**Comment by HowardHinnant** [comment]
:+1:

**Comment by rec** [comment]
Squashed into a minimal number of commits...

I hit a JUMP but all seems well now...

**Comment by HowardHinnant** [comment]
:+1:

**Comment by vinniefalco** [comment]
`src\ripple/rpc/impl/Utilities.h(34): warning C4099: 'ripple::RPC::Context' : type name first seen using 'struct' now seen using 'class' (..\..\src\ripple\unity\rpcx.cpp)`

**Comment by vinniefalco** [comment]
Non-unity fails to build in MSVC. Reason: No two translation units may have the same name, even in different directories (Ledger.cpp for example).

**Comment by HowardHinnant** [comment]
I did not know that!  That actually enforces what I believe to be a pretty good policy anyway.

**Comment by rec** [comment]
Fixed these two issues.

**Comment by rec** [comment]
Closing in favor of https://github.com/ripple/rippled/pull/1181

**Comment by rec** [comment]
All right!  I just rebased this against develop.  Fire away!

**Pull request merged**